File: all/python3/539.最小时间差.py
#
# @lc app=leetcode.cn id=539 lang=python3
#
# [539] 最小时间差
#

# @lc code=start
from typing import List


class Solution:
    def findMinDifference(self, timePoints: List[str]) -> int:
        n = len(timePoints)
        # 抽屉原理（鸽巢理论）
        if n > 1440:
            return 0
        times = []
        for time in timePoints:
            h, m = int(time[:2]), int(time[-2:])
            times.append(
                h * 60 + m,
            )
            times.append(h * 60 + m + 1440)
        # print(times)
        times.sort()
        # print(times)
        return min(times[i] - times[i - 1] for i in range(1, 2 * n))


# @lc code=end

File: all/typescript/539.最小时间差.ts
/*
 * @lc app=leetcode.cn id=539 lang=typescript
 *
 * [539] 最小时间差
 */

// @lc code=start
function findMinDifference(timePoints: string[]): number {
  const ans = []
  for (const timePoint of timePoints) {
    const h = parseInt(timePoint.substring(0, 2))
    const m = parseInt(timePoint.substring(3))
    ans.push(h * 60 + m)
    ans.push(h * 60 + m + 24 * 60)
  }
  // console.log(ans)
  // Array.sort 影响数组本身，并且默认是按照字符串排序（字典序）
  ans.sort((a, b) => a - b)
  // console.log(ans)
  let min = Infinity
  for (let i = 1; i < ans.length; i++) {
    const t1 = ans[i]
    const t2 = ans[i - 1]
    min = Math.min(min, t1 - t2)
  }
  return min
}
// @lc code=end
